src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+#if DEBUG
+using System.Diagnostics;
+#endif
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        /// <summary>
+        /// A reference-counting wrapper which allows multiple uses of a single disposable object in code, which is
+        /// deterministically released (by calling <see cref="IDisposable.Dispose"/>) when the last reference is
+        /// disposed.
+        /// </summary>
+        /// <remarks>
+        /// <para>While each instance of <see cref="ReferenceCountedDisposable{T}"/> should be explicitly disposed when
+        /// the object is no longer needed by the code owning the reference, this implementation will not leak resources
+        /// in the event one or more callers fail to do so. The underlying object will be deterministically released
+        /// when the last reference to it is disposed. However, the underlying object is eligible for non-deterministic
+        /// release (should it have a finalizer) when each reference to it is <em>either</em> disposed <em>or</em>
+        /// eligible for garbage collection itself.</para>
+        ///
+        /// <para>All public methods on this type adhere to their pre- and post-conditions and will not invalidate state
+        /// even in concurrent execution.</para>
```

Comment:
➡️ Don't blame me. Y'all are the ones who assigned me to the same team as @jaredpar.